embedded-graphics/src/draw_target.rs

```diff
@@ -222,161 +268,136 @@ where
-    /// internals of the [`draw`] method used for the [`Styled`] [`Triangle`] primitive. To draw a
-    /// triangle, call [`draw`] on a `Styled<Triangle>` object.
+    /// struct ExampleDisplay;
     ///
-    /// [`Triangle`]: ../primitives/triangle/struct.Triangle.html
-    /// [`draw`]: ./trait.DrawTarget.html#method.draw
-    /// [`Styled`]: ../style/struct.Styled.html
-    fn draw_triangle(
-        &mut self,
-        item: &Styled<primitives::Triangle, PrimitiveStyle<C>>,
-    ) -> Result<(), Self::Error> {
-        self.draw_iter(item)
-    }
-
-    /// Draws a styled rectangle primitive.
+    /// impl DrawTarget for ExampleDisplay {
+    ///     type Color = Gray8;
+    ///     type Error = core::convert::Infallible;
     ///
-    /// This default trait method can be overridden if a display provides hardware-accelerated
-    /// methods for drawing rectangle.
+    ///     fn size(&self) -> Size {
+    ///         Size::new(64, 64)
+    ///     }
     ///
-    /// # Caution
-    ///
-    /// This method should not be called directly from application code. It is used to define the
-    /// internals of the [`draw`] method used for the [`Styled`] [`Rectangle`] primitive. To draw a
-    /// rectangle, call [`draw`] on a `Styled<Rectangle>` object.
-    ///
-    /// [`Rectangle`]: ../primitives/rectangle/struct.Rectangle.html
-    /// [`draw`]: ./trait.DrawTarget.html#method.draw
-    /// [`Styled`]: ../style/struct.Styled.html
-    fn draw_rectangle(
-        &mut self,
-        item: &Styled<primitives::Rectangle, PrimitiveStyle<C>>,
-    ) -> Result<(), Self::Error> {
-        self.draw_iter(item)
-    }
-
-    /// Draws a styled circle primitive.
+    ///     fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+    ///     where
+    ///         I: IntoIterator<Item = Pixel<Self::Color>>,
+    ///     {
+    ///         // Draw pixels to the display
     ///
-    /// This default trait method can be overridden if a display provides hardware-accelerated
-    /// methods for drawing circles.
+    ///         Ok(())
+    ///     }
     ///
-    /// # Caution
+    ///     fn fill_contiguous<I>(&mut self, area: &Rectangle, colors: I) -> Result<(), Self::Error>
+    ///     where
+    ///         I: IntoIterator<Item = Self::Color>,
+    ///     {
+    ///         // Clamp area to drawable part of the display target
+    ///         let area = area.intersection(&Rectangle::new(Point::zero(), self.size()));
     ///
-    /// This method should not be called directly from application code. It is used to define the
-    /// internals of the [`draw`] method used for the [`Styled`] [`Circle`] primitive. To draw a
-    /// circle, call [`draw`] on a `Styled<Circle>` object.
+    ///         self.draw_iter(
+    ///             area.points()
+    ///                 .zip(colors)
+    ///                 .map(|(pos, color)| Pixel(pos, color)),
+    ///         )
+    ///     }
```

Comment:
I just noticed that this won't work. The intersection might, for example, be narrower than `area`, which means that each row of the points iter has fewer pixels than the colors iter which will mess up the display.

Reply:
Oh nuts, very true :facepalm:. We could do this:

```rust
// Clamp area to drawable part of the display target
let drawable_area = area.intersection(&Rectangle::new(Point::zero(), self.size()));

self.draw_iter(
    area.points()
        .zip(colors)
        .filter(|(pos, _color)| drawable_area.contains(pos))
        .map(|(pos, color)| Pixel(pos, color)),
)
```

Although it's pretty inefficient for both the demo code and the default impl. Not sure there's another way round it tbh.

Reply:
> Although it's pretty inefficient for both the demo code and the default impl.

The default impl doesn't need to clip, so it's fine the way it is.

For the demo code we could go with the example you provided. It's easy to understand and shouldn't be too bad performance wise. If we want to improve this in the future we could implement some kind of adapter around the iterator that skips the appropriate amount of pixels. But that seems more complicated then necessary at the moment.

Reply:
> For the demo code we could go with the example you provided.

Changes pushed :ok_hand: